packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,409 @@
+    );
+  }
+
+  async updateCurrentBundleAndDetails(
+    height: number,
+  ): Promise<UnZippedKyveBlockReponse[]> {
+    // this is on init, and then when height is greater than current cache
+    if (
+      this.currentBundleId === -1 ||
+      parseDecimal(this.cachedBundleDetails.to_key) < height
+    ) {
+      this.currentBundleId =
+        this.currentBundleId === -1
+          ? await this.getBundleId(height)
+          : this.currentBundleId + 1;
+      this.cachedBundleDetails = await this.getBundleById(this.currentBundleId);
+      try {
+        await fs.promises.access(
+          path.join(
+            this.tmpCacheDir,
+            `bundle_${parseDecimal(this.cachedBundleDetails.id) - 2}`,
+          ),
+        );
+        await this.clearFileCache();
+      } catch (e) {
+        /* if file does not exist, no need to clear fileCache */
+      }
+
+      return JSON.parse(await this.getFileCacheData());
+    } else {
+      return JSON.parse(
+        await this.readFromFile(
+          this.getBundleFilePath(this.cachedBundleDetails.id),
+        ),
+      );
+    }
+  }
+
+  private async pollUntilReadable(bundleFilePath: string): Promise<string> {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      try {
+        return await this.readFromFile(bundleFilePath);
+      } catch (e) {
+        if (e.code === 'EACCES') {
+          await delay(POLL_TIMER);
+        } else {
+          throw e;
+        }
+      }
+    }
+  }
+
+  async downloadAndProcessBundle(bundleFilePath: string): Promise<void> {
+    const writeStream = fs.createWriteStream(bundleFilePath, {
+      flags: 'wx',
+      mode: 0o200, // write only access for owner
+    });
+
+    await new Promise((resolve, reject) => {
+      writeStream.on('open', resolve);
+      writeStream.on('error', (err) => {
+        reject(err);
+      });
+    }).catch((e) => {
+      throw e;
+    });
+
+    const zippedBundleData = await this.retrieveBundleData();
+
+    const gunzip = zlib.createUnzip({
+      maxOutputLength: MAX_COMPRESSION_BYTE_SIZE /* avoid zip bombs */,
+    });
+
+    await new Promise((resolve, reject) => {
+      zippedBundleData.data
+        .on('error', (err) => reject(err))
+        .pipe(gunzip)
+        .on('error', (err) => reject(err))
+        .pipe(writeStream)
+        .on('finish', resolve);
+    }).catch((e) => {
+      throw e;
+    });
+
+    await fs.promises.chmod(bundleFilePath, 0o444);
+  }
+
+  async getFileCacheData(): Promise<string> {
+    const bundleFilePath = this.getBundleFilePath(this.cachedBundleDetails.id);
+
+    try {
+      await this.downloadAndProcessBundle(bundleFilePath);
+      return await this.readFromFile(bundleFilePath);
+    } catch (e: any) {
+      if (['EEXIST', 'EACCES', 'ENOENT'].includes(e.code)) {
+        return this.pollUntilReadable(bundleFilePath);
+      } else {
+        await fs.promises.unlink(bundleFilePath);
```

Comment:
If `this.readFromFile` throws an error, it will result in the file being deleted